testing/src/main/java/org/oppia/android/testing/media/TestMediaPlayerRobolectricImpl.kt

```diff
@@ -0,0 +1,169 @@
+  }
+
+  /** Calls Robolectric's Shadows.shadowOf() using reflection. */
+  private fun shadowOf(mediaPlayer: MediaPlayer): Any? {
+    val shadowsClass = getShadowsClass()
+
+    return shadowsClass.getMethod("shadowOf", MediaPlayer::class.java)
+      .invoke(/* obj= */ null, mediaPlayer)
+  }
+
+  /** Calls ShadowMediaPlayer.setDataSource() using reflection. */
+  private fun setDataSource(shadowMediaPlayer: Any, dataSource: Any) {
+    val dataSourceClass = getDataSourceClass()
+
+    shadowMediaPlayer.javaClass.getMethod("setDataSource", dataSourceClass)
+      .invoke(shadowMediaPlayer, dataSource)
+  }
+
+  /** Calls ShadowMediaPlayer.invokePreparedListener() using reflection. */
+  private fun invokePreparedListener(shadowMediaPlayer: Any) {
+    shadowMediaPlayer.javaClass.getMethod("invokePreparedListener").invoke(shadowMediaPlayer)
+  }
+
+  override fun setupAudio(explorationId: String, audioFileName: String) {
+    val dataSource = createAudioDataSource(
+      explorationId = explorationId, audioFileName = audioFileName
+    )
+    addShadowMediaPlayerException(dataSource!!, IOException("Test does not have networking"))
+  }
+
+  override fun setupAudioDualMedia(
+    explorationId1: String,
+    audioFileName1: String,
+    explorationId2: String,
+    audioFileName2: String
+  ) {
+    val dataSource = createAudioDataSource(
+      explorationId = explorationId1, audioFileName = audioFileName1
+    )
+    val dataSource2 = createAudioDataSource(
+      explorationId = explorationId2, audioFileName = audioFileName2
+    )
+    addShadowMediaPlayerException(dataSource!!, IOException("Test does not have networking"))
+    addShadowMediaPlayerException(dataSource2!!, IOException("Test does not have networking"))
+  }
+
+  @Suppress("SameParameterValue")
+  private fun createAudioDataSource(explorationId: String, audioFileName: String): Any? {
+    val audioUrl = createAudioUrl(explorationId, audioFileName)
+
+    val dataSourceClass = getDataSourceClass()
+
+    val toDataSource =
+      dataSourceClass.getDeclaredMethod(
+        "toDataSource", String::class.java, Map::class.java
+      )
+    return toDataSource.invoke(/* obj= */ null, audioUrl, /* headers= */ null)
+  }
+
+  private fun addShadowMediaPlayerException(dataSource: Any, exception: Exception) {
+    val shadowMediaPlayerClass = getShadowMediaPlayerClass()
+
+    val addException =
+      shadowMediaPlayerClass.getDeclaredMethod(
+        "addException", dataSource.javaClass, IOException::class.java
+      )
+    addException.invoke(/* obj= */ null, dataSource, exception)
+  }
+
+  private fun createAudioUrl(explorationId: String, audioFileName: String): String {
+    return "https://storage.googleapis.com/oppiaserver-resources/" +
+      "exploration/$explorationId/assets/audio/$audioFileName"
+  }
+
+  private fun getDataSourceClass(): Class<*> {
+    return Class.forName("org.robolectric.shadows.util.DataSource")
```

Comment:
Let's use the injected context's class loader rather than the root for this. It's generally cleaner to use classes closer to your test for retrieving classes since otherwise you can unintentionally bypass overwritten class loaders if you're not careful.

Reply:
Done

Reply:
resolving on behalf of Ben